feat(*): implement more granular pending task tracking

gkalpak opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Refactoring/Internal feature.


**What is the current behavior? (You can also link to an open issue here)**
Previously, all pending async tasks (tracked via `$browser`) are treated the same. I.e. things like `$$testability.whenStable()` and `ngMock#$timeout.verifyNoPendingTasks()` take all tasks into account.

Yet, in some cases we might be interested in specific tasks only. For example, if one wants to verify there are no pending `$timeout`s, they don't care if there are other pending tasks, such as `$http` requests. Similarly, one might want to get notified when all `$http` requests have completed and does not care about pending promises.


**What is the new behavior (if this is a feature change)?**
This PR adds support for more granular task tracking, by enabling callers to specify the type of task that is being added/removed from the queue and enabling listeners to be triggered when specific types of tasks are completed (even if there are more pending tasks of different types).

The change is backwards compatible. I.e. calling the affected methods with no explicit task-type, behaves the same as before.

_**UPDATE:** This PR now implements the API suggested in https://github.com/angular/angular.js/pull/16603#issuecomment-397797253._


**Does this PR introduce a breaking change?**
No.


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [x] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:
Closes #14336.

gkalpak committed:
feat(*): implement more granular pending task tracking

Previously, all pending async tasks (tracked via `$browser`) are treated
the same. I.e. things like `$$testability.whenStable()` and
`ngMock#$timeout.verifyNoPendingTasks()` take all tasks into account.

Yet, in some cases we might be interested in specific tasks only. For
example, if one wants to verify there are no pending `$timeout`s, they
don't care if there are other pending tasks, such as `$http` requests.
Similarly, one might want to get notified when all `$http` requests have
completed and does not care about pending promises.

This commit adds support for more granular task tracking, by enabling
callers to specify the type of task that is being added/removed from the
queue and enabling listeners to be triggered when specific types of
tasks are completed (even if there are more pending tasks of different
types).

The change is backwards compatible. I.e. calling the affected methods
with no explicit task-type, behaves the same as before.

Related to #14336.

gkalpak commented:
Some work is still needed, but I would like confirm we are itnerested in this change, before investing more time.

TODO:

- [ ] Add tests for changes in `ngMock`.
- [ ] _Optional_ Share task-tracking code between `$browser` and `ngMock.$browser` (e.g. by extracting the relevant bits into a reusable factory). Right now, the functionality is replicated in two places with no difference.
- [ ] _Optional_ Share code between `$interval` and `ngMock.$interval` (e.g. by extracting the relevant bits into a reusable factory). Right now the functionality is replicated in two places with minimal differences (e.g. `setTimeout`, `clearTimeout`).

##
In its current state, the PR does not introduce any change in external behavior, but adds internal features that could be built on top of. Here are my suggested public changes (in an attempts to address #14336 in a backwards compatible way): 

- [ ] **New API**: `ngMock.flushPendingTasks()`
    - _Alias for `ngMock.$timeout.flush()` and thus `$browser.defer.flush()`._
- [ ] **New API**: `ngMock.verifyNoPendingTasks([taskType])`
    - _Verifies that there are no pending tasks at all or no pending tasks of the specified type._
    - _Alias for `ngMock.$timeout.verifyNoPendingTasks()` (for historical reasons) when called with
     no argument._
    - _Additionally supports more granular checks with `taskType`._
- [ ] **Existing API**: `ngMock.$timeout.flush()`
    - _Leave unchanged (for backwards compatibility)._
    - _Document that all types of tasks (not just timeouts) are flushed._
    - _Recommend using `ngMock.flushPendingTasks()` instead. (It does the same, but is named more intuitively.)_
- [ ] **Existing API**: `ngMock.$timeout.verifyNoPendingTasks()`
    - _Leave unchanged (for backwards compatibility)._
    - _Document that all types of tasks (not just timeouts) are verified._
    - _Recommend using `ngMock.verifyNoPendingTasks('$timeout')` instead._

Narretz commented:
This sound like a good idea.
It could also help with e2e tests, provided protractor includes the changes to $$testability in the public API for waitForAngularEnabled() (see https://github.com/angular/angular.js/issues/14118). This is also where I see a problem: the proposed API allows whenStable to be called with a single argument, but I think for the most-wanted use case (ignoring timeouts that are used as intervals), you'd want to exclude a taskType rather than specifying one explicitly. This is especially important as many users may not be aware of stuff that uses applyAsync or evalAsync. I guess protractor could wrap multiple whenStable() executions, but I think we can have a better API here. This might also be an opportunity to make whenStable public.

thorn0 commented:
BTW, there remains one more unsolved problem with those tasks, namely timeouts and intervals have separate queues that have to be flushed separately. That's why it's difficult to unit-test logic that mixes timeouts and intervals. Another mention of this issue: https://github.com/angular/angular.js/pull/10525#issuecomment-69540468

gkalpak commented:
Tackling that would be a much bigger change (and one we probably won't get to :grin:).
I am not even sure how I feel about unifying the queues for `$timeout` and `$interval`, since they use different methods to schedule tasks (`setTimeout` vs `setInterval`) and you can't really guarrantee their relative order afaict (so why would tests make such guarrantees).

jbedard commented on src/ng/browser.js:
```diff
@@ -60,6 +73,33 @@ function Browser(window, document, $log, $sniffer) {
+  function decOutstandingRequestCount(taskType) {
+    taskType = taskType || DEFAULT_TASK_TYPE;
+    if (outstandingRequestCounts[taskType]) outstandingRequestCounts[taskType]--;
+    if (outstandingRequestCounts[ALL_TASKS_TYPE]) outstandingRequestCounts[ALL_TASKS_TYPE]--;
```
I feel like this should go into the `if (outstandingRequestCounts[taskType])` and not be it's own `if`. That way if `taskType` is invalid or not set it won't decrement `outstandingRequestCounts[ALL_TASKS_TYPE]`...?

gkalpak committed:
fixup! feat(*): implement more granular pending task tracking

gkalpak committed:
refactor(ngMock/$interval): more closely follow actual `$interval`'s internal implementation

gkalpak committed:
refactor($interval): share code between `$interval` and `ngMock/$interval`

This avoids code/logic duplication and helps the implementations stay
in-sync.

gkalpak committed:
refactor($browser): share task-tracking code between `$browser` and `ngMock/$browser`

This avoids code/logic duplication and helps the implementations stay
in-sync.

gkalpak committed:
feat(ngMock): add `$flushPendingTasks()` and `$verifyNoPendingTasks()`

`$flushPendingTasks([delay])` allows flushing all pending tasks (or up
to a specific delay). This includes `$timeout`s, `$q` promises and tasks
scheduled via `$rootScope.$applyAsync()` and `$rootScope.$evalAsync()`.
(ATM, it only flushes tasks scheduled via `$browser.defer()`, which does
not include `$http` requests and `$route` transitions.)

`$verifyNoPendingTasks([taskType])` allows verifying that there are no
pending tasks (in general or of a specific type). This includes tasks
flushed by `$flushPendingTasks()` as well as pending `$http` requests
and in-progress `$route` transitions.

Background:
`ngMock/$timeout` has `flush()` and `verifyNoPendingTasks()` methods,
but they take all kinds of tasks into account which is confusing. For
example, `$timeout.verifyNoPendingTasks()` can fail (even if there are
no pending timeouts) because of an unrelated pending `$http` request.

This behavior is retained for backwards compatibility, but the new
methods are more generic (and thus less confusing) and also allow
more fine-grained control (when appropriate).

Closes #14336

gkalpak committed:
docs(ngMock/$timeout): recommend `$verifyNoPendingTasks()` when appropriate

For historical reasons, `$timeout.verifyNoPendingTasks()` throws if
there is any type of task pending (even if it is not a timeout). When
calling `$timeout.verifyNoPendingTasks()`, the user is most likely
interested in verifying pending timeouts only, which is now possible
with `$verifyNoPendingTasks('$timeout')`.

To raise awareness of `$verifyNoPendingTasks()`, it is mentioned in the
error message thrown by `$timeoutverifyNoPendingTasks()` if none of the
pending tasks is a timeout.

gkalpak committed:
docs(ngMock/$timeout): deprecate `flush()` and `verifyNoPendingTasks()`

gkalpak committed:
fixup! feat(*): implement more granular pending task tracking

Narretz commented on src/ngMock/angular-mocks.js:
```diff
@@ -193,6 +239,72 @@ angular.mock.$Browser.prototype = {
+ *   account by `$verifyNoPendingTasks()`.
+ * </div>
+ *
+ * @param {string=} taskType - The type of tasks to check for.
```
`@returns` is missing. Should also include that it throws when tasks are pending.

gkalpak replied:
It doesn't really return anything atm :grin:
+1 for documenting the thrown error.

Narretz commented on src/ngMock/angular-mocks.js:
```diff
@@ -193,6 +239,72 @@ angular.mock.$Browser.prototype = {
+ *   These include tasks scheduled via `$evalAsync()` indirectly (such as {@link $q} promises).
+ *
+ * <div class="alert alert-info">
+ *   Periodic tasks scheduled via {@link $interval} use a different queue and are not taken into
```
Should we mention that it's impossible to verify that no intervals are running? "uses a different queue" makes it sound a bit like the possibility exists.

Narretz commented on src/ngMock/angular-mocks.js:
```diff
@@ -193,6 +239,72 @@ angular.mock.$Browser.prototype = {
+ * The types of tasks that are flushed include:
+ *
+ * - Pending timeouts (via {@link $timeout}).
+ * - Pending tasks scheduled via {@link ng.$rootScope.Scope#$applyAsync}.
```
This must be `{@link ng.$rootScope.Scope#$applyAsync $applyAsync}` or similar, otherwise it'll show `$rootScope.Scope` as the name. 

Same in the line below.

Narretz commented on src/ng/interval.js:
```diff
@@ -4,10 +4,18 @@ var $intervalMinErr = minErr('$interval');
 function $IntervalProvider() {
-  this.$get = ['$rootScope', '$window', '$q', '$$q', '$browser',
-       function($rootScope,   $window,   $q,   $$q,   $browser) {
+  this.$get = ['$$intervalFactory', '$window',
```
I assume we don't care that this is "breaking" the decoration of $interval by including a private service?

gkalpak replied:
Why does it break it?

Narretz replied:
Now that you ask ... doesn't look like it breaks anything (and it relied on a private service before)

Narretz commented on src/ngMock/angular-mocks.js:
```diff
@@ -2180,39 +2268,81 @@ angular.mock.$TimeoutDecorator = ['$delegate', '$browser', function($delegate, $
+          'If you only want to verify pending timeouts, use ' +
+          '`$verifyNoPendingTasks(\'$timeout\')` instead.';
+
+      throw new Error('Deferred tasks to flush (' + pendingTasks.length + '):\n  ' +
```
I think we should document that this fn will throw this error. I know that it's not new, but since we're at it ...

Narretz commented on src/ngMock/angular-mocks.js:
```diff
@@ -135,26 +131,76 @@ angular.mock.$Browser = function() {
+   * Verifies that there are no pending tasks that need to be flushed.
+   * You can check for a specific type of tasks only, by specifying a `taskType`.
+   *
+   * See {@link $verifyNoPendingsTasks} for more info.
```
*verifyNoPending**s**Tasks* -> *verifyNoPendingTasks*

why does this not fail in the docs processor?

Narretz reviewed (changes requested):
That's a lot of code change ...
I can't say I reviewed everything - mostly focussed on the docs.

So the second parameter from whenStable was removed intentionally because the API would have been too complex, right? I guess Protractor could still use the 2nd argument to `browser.notifyWhenNoOutstandingRequests()` to ignore some task types ($timeout, cough) on startup. Both browser and whenStable are private APIs anyway. 
I guess at this point it doesn't make much sense to make them public.

gkalpak committed:
fixup! feat(ngMock): add `$flushPendingTasks()` and `$verifyNoPendingTasks()`

gkalpak commented on src/ngMock/angular-mocks.js:
```diff
@@ -193,6 +239,72 @@ angular.mock.$Browser.prototype = {
+ *   `$flushPendingTasks()`. Use {@link ngMock.$interval#flush $interval.flush([millis])} instead.
+ * </div>
+ *
+ * @param {number=} delay - The number of milliseconds to flush.
```
What do you mean? It doesn't apply to `$timeout` only (I ahve listed the types of tasks flushed above).

Narretz replied:
I mean "the number milliseconds to flush" only applies to the $timeout queue. This might have been obvious when flush() was on $timeout, but now the flush() also flushes applyAsync and evalAsync and they don't have a "timeline" like timeout does. At least I would expect if I put in a delay of 0 or no delay that applyAsync and evalAsync would still be flushed, but timeouts wouldn't

gkalpak replied:
Good point. I'll mention that. I will ensure there are tests that verify the expected behavior (no delay flushes everything, 0 delay flushes stuff with 0 delay only, etc.).

Narretz commented:
> The reason we dropped the whenStable() changes (and kept it private 😁) is that Angular has since made different changes to their whenStable() API

Ok, but if you only think about AngularJS then Protractor could provide a different API for AngularJS apps. In fact, in the docs I can't really see anything related to timeouts passed into https://www.protractortest.org/#/api?view=ProtractorBrowser.prototype.waitForAngularEnabled 

However, they can still do it without the API being public.

gkalpak commented:
Yeah, protractor is still not using Angular's new `whenStable` API (but theoretically it will in the future).
And yes, protractor could probably use the private `$browser.notifyWhenNoPendingTasks()` API for AngularJS apps, if they want to, but it might be less straight forward than using `$$testability`, because we expose `$$testability` in an easier way iirc (whereas for `$browser`, they would have to get hold of the `$injector` first). (I could be wrong though.)

Narretz commented:
Protractor uses testability or browser via the injector: https://github.com/angular/protractor/blob/ed955e56a839d7f69da43acb6755763220d3681d/lib/clientsidescripts.js#L78-L87

Now if they wanted to use granular tracking they'd have to add some more conditions because granular tracking is only available via browser.

I think we should simply give whenStable the second argument back to make this easier.

gkalpak committed:
fixup! feat(ngMock): add `$flushPendingTasks()` and `$verifyNoPendingTasks()`

gkalpak commented:
We discussed this "offline" and decided that changing the testability API does not have to be part of this PR. Let's get this merged and discuss this with the Protractor team.

petebacondarwin commented on src/ng/browser.js:
```diff
@@ -22,61 +22,27 @@
-      outstandingRequestCallbacks.push(callback);
-    }
-  };
+  // TODO(vojta): prefix this method with $$ ?
```
It's probably a bit late for this TODO now :-)

gkalpak replied:
Hope never dies :smiley:

petebacondarwin commented on src/ng/taskTrackerFactory.js:
```diff
@@ -0,0 +1,122 @@
+ * A `TaskTracker` can keep track of pending tasks (grouped by type) and can notify interested
+ * parties when all pending tasks (or tasks of a specific type) have been completed.
+ *
+ * @param {$log} log - A logger instance (such as `$log`). Used to log error during callback
```
Doesn't this mean that the type of the `log` parameter is `$log`? Is that what you meant?

gkalpak replied:
Yes :grin: (Note: This is internal, non-public docs.)

petebacondarwin commented on src/ng/taskTrackerFactory.js:
```diff
@@ -0,0 +1,122 @@
+  var taskCallbacks = [];
+
+  var ALL_TASKS_TYPE = self.ALL_TASKS_TYPE = '$$all$$';
+  var DEFAULT_TASK_TYPE = self.DEFAULT_TASK_TYPE = '$$default$$';
```
Perhaps we should add a few more `$` signs here. Just to be safe :-P

gkalpak replied:
Don't be that guy ---> 🤑
(💲 is just a means to an end 😛)

petebacondarwin commented on src/ng/taskTrackerFactory.js:
```diff
@@ -0,0 +1,122 @@
+  /**
+   * Execute the specified callback when all pending tasks have been completed.
+   *
+   * If there are no pending tasks, the callback is executed immediatelly. You can optionally limit
```
typo: immediately

gkalpak replied:
😱

petebacondarwin commented on src/ng/taskTrackerFactory.js:
```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+/**
+ * ! This is a private undocumented service !
```
LOL - I can see the documentation below! :-P

gkalpak replied:
:grin: (I copied it from the docs of another private "undocumented" service 😛)

petebacondarwin commented on src/ng/taskTrackerFactory.js:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  function notifyWhenNoPendingTasks(callback, taskType) {
+    taskType = taskType || ALL_TASKS_TYPE;
```
I note that here we default to `ALL_TASKS_TYPE`, which makes me wonder if there is any real need for the `DEFAULT_TASKS_TYPE`, since the "user" would not be able to run `notifyWhenNoPendingTasks(cb, DEFAULT_TASKS_TYPE)` and so there is no point in tracking that category of tasks separately from the `ALL_TASKS_TYPE`.

gkalpak replied:
We could refactor the code in several different ways, but:

a. Consistency FTW
b. This ship has been merg...ehm...sailed 😛
c. Mostly consistency FTW 😇

gkalpak replied:
Oh...I was pretty sure this had been merged. Apparently not 🤔

petebacondarwin commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -614,47 +614,292 @@ describe('ngMock', function() {
-      expect(log).toEqual('A;');
+        browser.defer.cancel(taskId1);
+        expect(function() {browser.defer.verifyNoPendingTasks('fooType');}).not.toThrow();
+        expect(function() {browser.defer.verifyNoPendingTasks();}).toThrow();
```
should we check the type of the error that is thrown?

gkalpak replied:
This test focuses on `defer.cancel()`. There are tests in `defer.verifyNoPendingTasks()` that have assertions against the actual error message.

petebacondarwin reviewed (approved):
Generally LGTM - a few typos to fix and a couple of questions to answer.

gkalpak commented on test/ngMock/angular-mocksSpec.js:
```diff
@@ -614,47 +614,292 @@ describe('ngMock', function() {
+        callback = jasmine.createSpy('callback');
+      });
+
+      it('should immediatelly run the callback if no pending tasks', function() {
```
😱

gkalpak committed:
fixup! feat(*): implement more granular pending task tracking

gkalpak committed:
fixup! refactor($browser): share task-tracking code between `$browser` and `ngMock/$browser`

gkalpak commented:
Thx for the review, @petebacondarwin. I think I've addressed all feedback (either by making the changes or trying to be entertaining 😛).

petebacondarwin commented:
You were just about entertaining enough to distract me from requesting you make any more changes. Which I think means I am happy to let you merge this.

Narretz reviewed (approved):
🚀
